File: 프로그래머스/lv2/92341. 주차 요금 계산/주차 요금 계산.cpp
#include <string>
#include <vector>
#include <map>
#include <unordered_map>

using namespace std;

int convert(string time)
{
    string hh = time.substr(0,2);
    string mm = time.substr(3);
    return stoi(hh) *60 + stoi(mm);
}

vector<int> solution(vector<int> fees, vector<string> records) {
    unordered_map<string, int> intime;
    map<string, int> price;
    
    for (string r: records)
    {
        string time = r.substr(0,5);
        string car_num = r.substr(6,4);
        string status = r.substr(11);
        if ( status =="IN")
        {
            intime[car_num] = convert(time);
        }
        else
        {
            price[car_num] += convert(time) - intime[car_num];
            intime.erase(car_num);
        }
    }
    
    for ( auto k : intime)
    {
        price[k.first] += 23*60+59 - k.second;
    }
    
    vector<int> answer;
    for (auto item : price)
    {
        if (item.second <= fees[0])
        {
            answer.push_back(fees[1]);
        }
        else
        {
            int temp = (item.second-fees[0]) / fees[2];
            if ((item.second-fees[0]) % fees[2] != 0)
            {
                temp+=1;
            }
            answer.push_back(fees[1] + temp*fees[3]);
        }
    }
    return answer;
}

File: 프로그래머스/lv2/92341. 주차 요금 계산/주차 요금 계산.py
def money(fees,time):
    import math
    result = []
    car_num = sorted(time)
    for num in car_num:
        if (time[num] <= fees[0]):
            result.append(fees[1])
        else:
            tmp = fees[1] + math.ceil((time[num]-fees[0]) / fees[2]) * fees[3]
            result.append(tmp)
    return result
def time_count(intime,outtime):
    intime_min = int(intime[:2]) * 60 + int(intime[3:])
    outtime_min = int(outtime[:2]) * 60 + int(outtime[3:])
    time_count = outtime_min - intime_min
    return time_count
def solution(fees, records):
    answer = []
    parking_time = {num.split(" ")[1] : 0 for num in records}
    parking = {}
    for record in records:
        tmp = record.split(" ")
        if tmp[2] == "IN":
            parking[tmp[1]] = tmp[0]
        elif tmp[2] =="OUT":
            parking_time[tmp[1]] += time_count(parking[tmp[1]],tmp[0])
            parking.pop(tmp[1])
    if parking:
        for cars in parking.keys():
            parking_time[cars]+=time_count(parking[cars],"23:59")
    answer = money(fees,parking_time)
    return answer